rcl/test/rcl/test_guard_condition.cpp

```diff
@@ -0,0 +1,201 @@
+  EXPECT_EQ(nullptr, gc_handle);
+  rcl_reset_error();
+  gc_handle = rcl_guard_condition_get_rmw_handle(&zero_guard_condition);
+  EXPECT_EQ(nullptr, gc_handle);
+  rcl_reset_error();
+  start_memory_checking();
+  assert_no_malloc_begin();
+  assert_no_realloc_begin();
+  assert_no_free_begin();
+  gc_handle = rcl_guard_condition_get_rmw_handle(&guard_condition);
+  assert_no_malloc_end();
+  assert_no_realloc_end();
+  assert_no_free_end();
+  stop_memory_checking();
+  EXPECT_NE(nullptr, gc_handle);
+}
+
+/* Tests the guard condition life cycle, including rcl_guard_condition_init/fini().
+ */
+TEST_F(
+  CLASSNAME(TestGuardConditionFixture, RMW_IMPLEMENTATION), test_rcl_guard_condition_life_cycle) {
+  stop_memory_checking();
+  rcl_ret_t ret;
+  rcl_guard_condition_t guard_condition = rcl_get_zero_initialized_guard_condition();
+  rcl_guard_condition_options_t default_options = rcl_guard_condition_get_default_options();
+  // Trying to init before rcl_init() should fail.
+  ret = rcl_guard_condition_init(&guard_condition, default_options);
+  ASSERT_EQ(RCL_RET_NOT_INIT, ret) << "Expected RCL_RET_NOT_INIT";
+  ASSERT_TRUE(rcl_error_is_set());
+  rcl_reset_error();
+  // Initialize rcl with rcl_init().
+  ret = rcl_init(0, nullptr, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret);
+  auto rcl_shutdown_exit = make_scope_exit([]() {
+    rcl_ret_t ret = rcl_shutdown();
+    ASSERT_EQ(RCL_RET_OK, ret);
+  });
+  // Try invalid arguments.
+  ret = rcl_guard_condition_init(nullptr, default_options);
+  ASSERT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << "Expected RCL_RET_INVALID_ARGUMENT";
+  ASSERT_TRUE(rcl_error_is_set());
+  rcl_reset_error();
+  // Try with invalid allocator.
+  rcl_guard_condition_options_t options_with_invalid_allocator =
+    rcl_guard_condition_get_default_options();
+  options_with_invalid_allocator.allocator.allocate = nullptr;
+  options_with_invalid_allocator.allocator.deallocate = nullptr;
+  options_with_invalid_allocator.allocator.reallocate = nullptr;
+  ret = rcl_guard_condition_init(&guard_condition, options_with_invalid_allocator);
+  ASSERT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << "Expected RCL_RET_INVALID_ARGUMENT";
+  ASSERT_TRUE(rcl_error_is_set());
+  rcl_reset_error();
+  // Try with failing allocator.
+  rcl_guard_condition_options_t options_with_failing_allocator =
+    rcl_guard_condition_get_default_options();
+  options_with_failing_allocator.allocator.allocate = failing_malloc;
+  options_with_failing_allocator.allocator.deallocate = failing_free;
+  options_with_failing_allocator.allocator.reallocate = failing_realloc;
+  ret = rcl_guard_condition_init(&guard_condition, options_with_failing_allocator);
+  ASSERT_EQ(RCL_RET_BAD_ALLOC, ret) << "Expected RCL_RET_BAD_ALLOC";
+  // The error will not be set because the allocator will not work.
+  // It should, however, print a message to the screen and get the bad alloc ret code.
+  // ASSERT_TRUE(rcl_error_is_set());
+  // rcl_reset_error();
```

Comment:
Yeah, I explicitly kept those as an example of what I would normally do, but I am not due to the comment above.